man/nsmaint.8
.TH NSMAINT 8 local
.DA 09 June 1989
.SH NSMAINT
nsmaint \- allow non\-root users to maintain the name server and data base
.SH SYNOPSIS
.B nsmaint
[
.B -d
]
sync|kill|start|stop|dump|clean|START|...
.SH DESCRIPTION
.I Nsmaint
is a setuid program intended to allow non\-root users the ability
to stop and start the
.IR named (8)
name server daemon and to edit files in its database.  The program
assumes that all name server data base files are stored in a directory
tree rooted in a single directory (conventionally
.IR /etc/nsdata )
and that users who are name server maintainers will be members of a
group named
.IR nsmaint .
.PP
The program is invoked with a single keyword, which indicates the
operation to be performed, as an argument.  The following keywords
are understood.
.PP
.B sync
.PP
Causes the running daemon to reread its data base (by sending a
SIGHUP to the daemon).  This is useful for
making the name server aware of changes which have been made to its
data base files.
.PP
.B stop
.PP
Causes the running daemon to terminate by sending a SIGTERM signal to it.
.PP
.B kill
.PP
Currently a synonym for the
.I stop
keyword.
.PP
.B start
.PP
Causes a new instance of the daemon to be started.  This is equivalent
to running the named program from a root account except that
.I nsmaint
will first try to determine if a named is running already and, if so,
will politely decline to start a second copy.
.PP
.B START
.PP
Like
.IR start ,
except that the check for an already running daemon is not performed.  The
check for an existing daemon can occasionally be mistaken.
.I START
is available to force the issue.
.PP
.B dump
.PP
Sends a SIGINT signal to the running daemon, which should provoke it into
dumping its internal database into a disk file (conventionally
.IR /usr/tmp/named_dump.db ).
.PP
.B clean
.PP
Does a recursive cleanup of file ownerships and permissions in the
name server data base directory (conventionally
.IR /etc/nsdata ).
The ownership of all directories and singly linked regular files are
changed to user
.I root
and group 
.IR nsmaint ,
with permissions being set to 664 for regular files and 775 for
directories (in effect, all files are made group writable).
.I Nsmaint
will quietly decline to change the ownership or permissions of files
and directories with the sticky bit set.  It will also noisily decline to
change regular files with more than one link and any files which are
neither directories nor regular files, including symbolic links.  Attempts
to make
.I nsmaint
do something which it considers a possible attempt to breach security are
reported via
.IR syslog (3)
at high priority.  Note that while the root of the database tree (i.e.
.IR /etc/nsdata )
may be a symbolic link, links of any kind are disallowed below this point.
.PP
.B debuginc
.PP
Causes the debug level of the running daemon to be incremented, by sending
it a SIGUSR1 signal.
.PP
.B debugoff
.PP
Causes debugging to be turned off in the running daemon, by sending it a
SIGUSR2 signal.
.SH FILES
.nf
.ta \w'/usr/tmp/named_dump.db   'u
/etc/named	name server daemon executable
/etc/named.pid	the process id of the running daemon
/etc/nsdata	root of the database tree
.fi
.SH "SEE ALSO"
named(8)
.SH BUGS
.I Nsmaint
is a setuid program of considerable power.  It is better if members of
the nsmaint group are at least somewhat trusted users.


CHANGES
This is version 4.8.2 of bind.  It is a test release that includes
versions 4.8 and 4.8.1 with fixes.  Although it is currently described
as a test release, it is believed to be reasonably stable and more
usable than the currently-released versions 4.8 and 4.8.1.  Here are
some of the more important changes:

  o Support has been added for the the T_TXT data type and for the class
    C_HS.  These are both used by Hesiod from Project Athena at MIT.

  o All of the pathnames have been put into one header file.  This
    makes it easier to change the location based upon your local
    configuration.

  o Numerous bugs have been fixed:  Adding a new authoritative zone now
    works when the server has a cached SOA record.  Comparisons in the
    db now look at type and class as well instead of dropping records
    with identical data.  Scheduling of maintenance interrupts has been
    moved to one routine avoid spurious ones.  Named goes into the background
    after more of the initialization is done.  Stream connection queue
    handling was cleaned up including a bug that caused data corruption
    and core dumps.  Sys5 no longer can have multiple transfers of the
    same zone occuring at the same time.  Handle CNAME -> CNAME loops
    more gracefully.  Avoid making one server never get queried.

There are several bug reports that have yet to be integrated into this
version.  Hopefully they will be dealt with in the next release.  Please
send feedback on this release.

Notes from versions 4.8.1 and 4.8 follow:

------------------

This is version 4.8.1 of bind.  It is a test release that includes
version 4.8 with fixes, asynchronous zone transfer and better reload
capabilities.  Although it is currently described as a test release,
it is believed to be reasonably stable and more usable than the currently-
released version, 4.8.  The changes of note are:

  o The asynchronous zone transfer code previously posted to the bind
    mailing list has been integrated, completed and tested.  There are
    a number of changes from the version posted, including fixes to
    allow top-level domains to work and a simplification of the timer
    code.

  o The code for reloading the server has been changed so that only
    primary zones master files that have changed are reloaded.  The
    cache and secondary zones are not flushed, and the sortlist, domain,
    etc. are reset to correspond to the boot file contents.

  o Several bugs have been fixed: the name "*" is not interpreted as
    a wildcard in cached zones, only in primary zones.  Secondary servers
    no longer decrement the time-to-live of records by the time since
    they verified the zone with the master; as a result, they never
    hand out nameserver referrals with too short a TTL to be usable.
    A bug was fixed that caused secondary servers with out-of-date
    zones to return empty answers between the actual expiration time
    and the next timeout.

There are several other bugs that have been reported but have not yet
been fixed.  In addition, the next regular release of named will
support negative caching, but this has not been integrated.

I would appreciate receiving feedback on this release; in particular,
problems (or lack of problems) when installing on various systems.
I attempted to update the SysV code when integrating the zone-transfer,
but haven't tested it.

The notes from version 4.8 follow.

----------
Welcome to version 4.8 of bind.

There have been several changes to the named boot file (/etc/named.boot)
of which you should be aware.  The "domain" line for each zone is no longer
needed, but one such line may still be used to specify a default domain
to be used for queries containing names with only a single component.
The term "suffixes", which was added in version 4.7alpha, has been removed.

The manual page on named (named.8) has been updated to reflect all
these changes.  Please read this and look at the example files
before installation.  You should also note the changes in the
resolver code to support non-fully-qualified addresses and per-user
host aliases.  See hostname(7) for an overview.  Two new routines
have been added to the resolver library since the last test release:
res_query formulates a query, sends it, waits for a response and does
preliminary error checking; res_search implements the search rules
of gethostbyname using res_query.

The MX lookup routine in sendmail has been modified to use res_search.  
Also, dn_skip takes an additional parameter and has been renamed
to dn_skipname.  While old sendmail binaries will work with the new
version of bind, because of these changes, it is desirable to install
new sendmail sources and recompile sendmail.  Do not rebuild sendmail 
from old sources.   The new sendmail is on ucbarpa.Berkeley.EDU for
anonymous FTP from pub/4.3/sendmail.MX.tar and pub/4.3/sendmail.MX.tar.Z.

There have been numerous changes to named, fixing most of the known
bugs that can be fixed without major structural changes in the server.
Several server configurations that failed before should now work.
Certain robustness problems have been fixed, in particular bounds-
checking when processing incoming packets.  Two changes have been made
in preparation for negative caching: SOA records are sent in the authority
section in negative responses with NXDOMAIN set, and a bug was fixed that
caused confusion and repeated requests if a response had no error, no answer
and an SOA in the authority section.  As such responses are already sent
by other servers, and will be sent by the next release of BIND, it is
important that all sites upgrade to this version as quickly as possible.

The root "hint" cache and cache file remain the largest problem area,
along with named's naivete in accepting bogus server's data.
These will be addressed in the next release, along with asynchronous
zone transfers, intelligent reloading of zone files, faster startup,
and caching of negative responses.


tools/nsmaint.c
/*
 * nsmaint - allow privileged souls to diddle the name server
 *
 * by Dennis Ferguson.
 */

#include <stdio.h>
#include <ctype.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/param.h>
#include <sys/dir.h>
#include <pwd.h>
#include <grp.h>
#include <sys/signal.h>
#include <syslog.h>
#include <errno.h>
#include <pathnames.h>

#define	STREQ(a, b)	(*(a) == *(b) && strcmp((a), (b)) == 0)

/*
 * People who are allowed to play name server god should be
 * in this group.
 */
#define	NSGROUP		"nsmaint"

/*
 * Program name we must be run with.  Want to make sure run attempts
 * end up in the accounting file in recognizable form.
 */
#define	THISPROGNAME	"nsmaint"

/*
 * When cleaning the data directory, we make sure these mode bits
 * are set on files and directories.
 */
#define	FILEMODETOSET	0664

/*
 * Path names to files we need to know about
 */

/* the nameserver pid file: /etc/named.pid */
#ifndef	PIDFILE
#define	PIDFILE	_PATH_PIDFILE
#endif

/* the nameserver program: /etc/named */
#ifndef	NAMED
#define	NAMED	_PATH_NAMED
#endif

/* the nameserver data directory: /etc/nsdata/. */
#ifndef	DATADIR
#define	DATADIR	_PATH_DATADIR
#endif

/*
 * Umask to run named with
 */
#define	NSUMASK	002

/*
 * Sanity.  We don't believe named will ever run with a pid less than this.
 */
#define	MINNSPID	3

/*
 * Phony environment so named is run relatively clean
 */
char *namedenv[] = {
	"HOME=/",
	"SHELL=/bin/sh",
	"TERM=dumb",
	"USER=root",
	NULL
};

/*
 * Hack to detect 'start' command.  This must be dealt with differently
 * than the others.
 */

typedef enum {	Error,
		Hup = SIGHUP,		/* sorry... */
		Term = SIGTERM,
		Int = SIGINT,
		Stats = SIGIOT,
		Usr1 = SIGUSR1,
		Usr2 = SIGUSR2,
		Start, Clean, ForceStart } Action;

/*
 * Command table.
 */
struct nscommands {
	char *cmdname;
	Action cmdaction;
} nscmds[] = {
	{ "sync",	Hup	},
	{ "kill",	Term	},
	{ "start",	Start	},
	{ "stop",	Term	},
	{ "dump",	Int	},
	{ "debuginc",	Usr1	},
	{ "debugoff",	Usr2	},
	{ "clean",	Clean	},
	{ "START",	ForceStart },
	{ "stats",	Stats	},
	{ NULL,		Error	}
};

char *thisprogname = THISPROGNAME;
char *nsprog = NAMED;
int userokay = 0;
char *username;		/* name of user running program */
int nsmaintgrp;		/* gid of the nsmaint group */

char *progname;
int debug = 0;

extern int errno;

/*
 * main - parse arguments and handle options
 */
main(argc, argv)
int argc;
char *argv[];
{
	int c;
	int errflg = 0;
	Action docmd;
	int namedpid;
	extern int optind, errno;
	extern char *optarg;
	extern Action getcmd();

	progname = argv[0];
	while ((c = getopt(argc, argv, "d")) != EOF)
		switch (c) {
		case 'd':
			++debug;
			break;
		default:
			errflg++;
			break;
		}
	if (!errflg &&
	    (optind != argc-1 || (docmd = getcmd(argv[optind])) == Error))
		errflg++;
	if (errflg) {
		setlinebuf(stderr);
		(void) fprintf(stderr,
		    "Usage: %s [-d] {%s", thisprogname, nscmds[0].cmdname);
		for (c = 1; nscmds[c].cmdname != NULL; ++c)
			(void) fprintf(stderr, "|%s", nscmds[c].cmdname);
		(void) fprintf(stderr, "}\n");
		exit(2);
	}

#ifdef	LOG_AUTH
	openlog(thisprogname, LOG_NOWAIT, LOG_AUTH);
#else
#ifdef	LOG_NOWAIT
	openlog(thisprogname, LOG_NOWAIT);
#else
	openlog(thisprogname, 0);
#endif
#endif
	userokay = checkoutuser();
	if (!userokay)
		exit(0);	/* ignore plebes */
	checkprogname(progname);
	namedpid = getnamedpid();

	switch (docmd) {
	case Start:
	case ForceStart:
		/*
		 * If the command is not forcestart, make sure
		 * named not running.
		 */
		if (docmd == Start && namedpid >= MINNSPID
		    && isrunning(namedpid)) {
			(void) fprintf(stderr,
			    "%s: %s (pid %d) appears to be running already\n",
			    thisprogname, nsprog, namedpid);
			exit(1);
		}
		if (debug)
			(void) printf("exec'ing %s\n", nsprog);
		else {
			(void) umask(NSUMASK);
			execle(nsprog, nsprog, (char *)NULL, namedenv);
			/* whoops! */
			Perror("can't exec %s", nsprog, "");
			exit(1);
		}
		break;
	case Clean:
		(void) umask(NSUMASK);
		if (chfile(DATADIR, 0, nsmaintgrp, FILEMODETOSET))
			chdirectory(DATADIR, 0, nsmaintgrp, FILEMODETOSET);
		break;
	case Term:
	case Hup:
	case Int:
	case Stats:
	case Usr1:
	case Usr2:
		/*
		 * If we're here, we have a signal to deliver.
		 */
		if (namedpid < MINNSPID) {
			(void) fprintf(stderr,
				       "%s: %s is either missing or contains an unlikely value\n",
				       thisprogname, PIDFILE);
			exit(1);
		}
		if (!isrunning(namedpid)) {
			(void) fprintf(stderr,
				       "%s: %s (pid %d) isn't running\n",
				       thisprogname, nsprog, namedpid);
			exit(1);
		}
		if (debug)
			(void) printf("sending signal %d to pid %d\n",
					       (int)docmd, namedpid);
		else if (kill(namedpid, (int)docmd) != 0) {
			Perror("failed to deliver signal to %s (pid %d)",
			    nsprog, (char *)namedpid);
			exit(1);
		}
		break;
	}
	exit(0);
}


/*
 * getsig - return the signal corresponding to this command
 */
Action
getcmd(cmd)
	char *cmd;
{
	register struct nscommands *nsc;

	for (nsc = nscmds; nsc->cmdname != NULL; nsc++)
		if (nsc->cmdname == NULL || STREQ(cmd, nsc->cmdname))
			return nsc->cmdaction;
	return Error;
}


/*
 * checkprogname - check the program name we were run with
 */
checkprogname(name)
	char *name;
{
	register char *cp;
	int nlen, thislen;
	int err = 1;

	/*
	 * The last element in the path in argv[0] is compared against
	 * the required program name.  This check is grossly inadequate,
	 * but prevents the easiest method of running the program invisibly
	 * to the accounting (with the shell).
	 */
	nlen = strlen(name);
	thislen = strlen(thisprogname);

	if (nlen == thislen) {
		cp = name;
		err = 0;
	} else if (nlen > thislen) {
		cp = &name[nlen-thislen];
		if (*(cp-1) == '/')
			err = 0;
	}

	if (!err && STREQ(cp, thisprogname))
		return;
	
	(void) fprintf(stderr, "%s: must run progam with name of %s\n",
	    thisprogname, thisprogname);
	/* Complain loudly, at least until holes are fixed */
	syslog(LOG_ALERT, "user %s attempted to run %s with name %s",
	    username, thisprogname, name);
	exit(1);
}


/*
 * checkoutuser - see if this user is one of the privileged few
 */
int
checkoutuser(docmd)
	Action docmd;
{
	register int i;
	struct passwd *pw;
	struct group *gr;
	int uid;
	int ngroups, gidset[NGROUPS];

	uid = getuid();
	if ((gr = getgrnam(NSGROUP)) == NULL) {
		/*
		 * Super gross hack.  A side effect of this routine
		 * is to set nsmaintgrp to the proper gid.  If root
		 * is running this, however, we don't necessarily
		 * want to prevent him from doing things even if the
		 * group doesn't exist, except that he can't do a clean
		 * command if we don't know the group.  Sigh.
		 */
		if (uid != 0 || docmd == Clean) {
			(void) fprintf(stderr, "%s: group %s doesn't exist\n",
			    thisprogname, NSGROUP);
			exit(1);
		}
		nsmaintgrp = -1;
	} else
		nsmaintgrp = gr->gr_gid;

	if (uid == 0) {
		username = "root";
		/*
		 * Let root do what he wants.
		 */
		return 1;
	}

	if ((pw = getpwuid(uid)) == NULL) {
		(void) fprintf(stderr, "%s: username for your uid (%d) unknown",
		    thisprogname, uid);
		syslog(LOG_ALERT, "unknown user with uid %d tried to run %s",
		    uid, thisprogname);
		exit(1);
	}
	username = pw->pw_name;

	if (nsmaintgrp == getgid())
		goto okay;	/* okay */
	
	if ((ngroups = getgroups(NGROUPS, gidset)) == -1) {
		(void) fprintf(stderr, "%s: getgroups() failed: ",
		    thisprogname);
		perror("");
	}

	for (i = 0; i < ngroups; i++)
		if (nsmaintgrp == gidset[i])
			goto okay;
	(void)setuid(uid);
	syslog(LOG_ALERT, "user %s attempted to run %s", username,
	    thisprogname);
	return 0;

okay:
	(void)setreuid(0, 0);
	return 1;
}


/*
 * getnamedpid - get named's pid from disk file
 */
int
getnamedpid()
{
	register char *cp;
	int pid;
	char buf[10];
	FILE *fp;

	if ((fp = fopen(PIDFILE, "r")) == NULL)
		return 0;
	buf[0] = '\0';
	(void) fgets(buf, sizeof buf, fp);
	(void) fclose(fp);

	for (cp = buf; *cp != '\0' && *cp != '\n'; cp++)
		if (!isdigit(*cp))
			return 0;
	if (*cp != '\0')
		*cp = '\0';
	if (buf[0] == '\0')
		return 0;
	
	pid = atoi(buf);
	if (pid < MINNSPID || pid > 32767)
		return 0;
	return pid;
}


/*
 * isrunning - see if program with this pid is running
 */
int
isrunning(pid)
	int pid;
{
	if (kill(pid, 0) == -1) {
		/*
		 * ESRCH means process not running.  Otherwise, user
		 * was unprivileged but program is going.
		 */
		if (errno == ESRCH)
			return 0;
	}
	return 1;
}


/*
 * chdirectory - (recursively) change owner, group and perm bits on files
 *		 in a directory.
 */
chdirectory(dir, uid, gid, modebits)
	char *dir;
	int uid, gid, modebits;
{
	register DIR *dirp;
	register struct direct *dp;
	char savedir[1024];

	if (getwd(savedir) == 0) {
		(void) fprintf(stderr, "%s: can't get working directory: %s\n",
		    thisprogname, savedir);
		exit(1);
	}

	if (chdir(dir) == -1) {
		Perror("can't change directory to %s", dir, "");
		exit(1);
	}
	if ((dirp = opendir(".")) == NULL) {
		Perror("failed to open directory %s", dir, "");
		exit(1);
	}

	dp = readdir(dirp);
	dp = readdir(dirp); /* read "." and ".." */
	for (dp = readdir(dirp); dp != NULL; dp = readdir(dirp))
		if (chfile(dp->d_name, uid, gid, modebits))
			chdirectory(dp->d_name, uid, gid, modebits);
	closedir(dirp);

	if (chdir(savedir) == -1) {
		Perror("can't change directory to %s", savedir, "");
		exit(1);
	}
}


/*
 * chfile - cautiously change the mode and permissions of a file.  Return
 *	1 if file was a directory, 0 otherwise.
 */
int
chfile(name, uid, gid, modebits)
	char *name;
	int uid, gid, modebits;
{
	struct stat st;
	int isdir = 0;
	int okaytochange = 0;
	int mode;

	/*
	 * Act real suspicious.  We only change directories and regular
	 * files, the latter only if they have a single hard link.
	 * Complain about everything else, including symbolic links.
	 */
	if (lstat(name, &st) != 0) {
		Perror("can't stat %s", name, "");
		exit(1);
	}

	switch(st.st_mode & S_IFMT) {
	case S_IFDIR:
		okaytochange = 1;
		isdir = 1;
		break;
	case S_IFREG:
		if (st.st_mode & (S_ISUID|S_ISGID)) {
			syslog(LOG_ALERT,
		"user %s attempting to change mode of set[ug]id file %s",
			    username, name);
			(void) fprintf(stderr, "%s: file %s is set[ug]id\n",
			    thisprogname, name);
		} else if (st.st_nlink > 1) {
			syslog(LOG_ALERT,
		"user %s attempting to change mode of file %s, has %d links",
			    username, name, st.st_nlink);
			(void) fprintf(stderr, "%s: file %s has %d links\n",
			    thisprogname, name, st.st_nlink);
		} else if (st.st_mode & S_ISVTX) {
			if (debug)
				printf("%s: will not change file %s\n",
					    thisprogname, name);
		} else {
			okaytochange = 1;
		}
		break;
	default:
		syslog(LOG_ALERT,
		"user %s attempting to change mode of %s, current mode is 0%o",
		    username, name, st.st_mode);
		fprintf(stderr,
		    "%s: file %s is neither directory nor regular file\n",
		    thisprogname, name);
		break;
	}

	if (okaytochange) {
		/*
		 * Only do the system calls if we need to.
		 */
		if ((uid != -1 && st.st_uid != uid) ||
		    (gid != -1 && st.st_gid != gid)) {
			if (debug) {
				(void) printf("chown'ing %s to %d,%d\n",
				    name, uid, gid);
			} else if (chown(name, uid, gid) != 0) {
				Perror("can't chown %s", name, "");
			}
		}
		if ((st.st_mode & modebits) != modebits) {
			mode = (st.st_mode | modebits) & 07777;
			if (debug) {
				(void) printf("chmod'ing %s to mode 0%o\n",
				    name, mode);
			} else if (chmod(name, mode) != 0) {
				Perror("can't chmod %s", name, "");
			}
		}
	}
	return isdir;
}



/*
 * Perror - print an error message
 */
Perror(fmt, s1, s2)
	char *fmt;
	char *s1, *s2;
{
	int e = errno;

	(void) fprintf(stderr, "%s: ", thisprogname);
	(void) fprintf(stderr, fmt, s1, s2);
	(void) fputs(": ", stderr);
	errno = e;
	perror("");
}
